File: cpp/tests/test_catalog.cpp
#include "forensicator/catalog.hpp"
#include "forensicator/util.hpp"

#include <cassert>
#include <cstdio>
#include <cstdlib>
#include <filesystem>
#include <iostream>

using namespace forensicator;

int main() {
    namespace fs = std::filesystem;
    auto tmp = fs::temp_directory_path() / ("fc-test-cat-" + std::to_string(now_epoch()));
    fs::create_directories(tmp);
    auto db = tmp / "c.sqlite";

    std::string schema = find_schema_file({});
    if (schema.empty()) {
        std::cerr << "schema not found via search; FORENSICATOR_SCHEMA env should be set\n";
        return 2;
    }

    {
        Catalog cat;
        cat.open(db, schema, false);
        auto v = cat.upsert_volume("h1", "v1", "/mnt", os_name());
        assert(v.volume_id == 1);
        assert(v.hostname == "h1");
        assert(v.volume == "v1");
        auto got = cat.get_volume("h1", "v1");
        assert(got.has_value());
        assert(got->volume_id == 1);
        auto vs = cat.volumes();
        assert(vs.size() == 1);
        cat.set_meta("dataset_name", "test-ds");
        auto sv = cat.get_meta("schema_version");
        assert(sv.has_value() && *sv == "1");
    }

    // Re-open
    {
        Catalog cat;
        cat.open(db, schema, true);
        auto v = cat.get_volume("h1", "v1");
        assert(v.has_value());
        assert(*cat.get_meta("dataset_name") == "test-ds");
    }

    std::error_code ec;
    fs::remove_all(tmp, ec);
    std::cout << "test_catalog ok\n";
    return 0;
}

File: python/tests/test_catalog.py
import sqlite3

from forensicator.catalog import (
    catalog_meta,
    get_volume,
    open_catalog,
    set_catalog_meta,
    upsert_volume,
    volumes,
)


def test_open_creates_catalog(tmp_path):
    p = tmp_path / "c.sqlite"
    conn = open_catalog(str(p))
    try:
        assert p.exists()
        sv = catalog_meta(conn, "schema_version")
        assert sv == "1"
    finally:
        conn.close()


def test_meta_roundtrip(tmp_path):
    conn = open_catalog(str(tmp_path / "c.sqlite"))
    try:
        set_catalog_meta(conn, "k", "v")
        assert catalog_meta(conn, "k") == "v"
        set_catalog_meta(conn, "k", "v2")
        assert catalog_meta(conn, "k") == "v2"
    finally:
        conn.close()


def test_upsert_volume_idempotent(tmp_path):
    conn = open_catalog(str(tmp_path / "c.sqlite"))
    try:
        v1 = upsert_volume(conn, hostname="h", volume="v", mount_point="/m", os="darwin")
        v2 = upsert_volume(conn, hostname="h", volume="v", mount_point="/m2", os="darwin")
        assert v1["volume_id"] == v2["volume_id"]
        got = get_volume(conn, "h", "v")
        assert got["mount_point"] == "/m2"
        all_vols = volumes(conn)
        assert len(all_vols) == 1
    finally:
        conn.close()


def test_schema_tables_created(tmp_path):
    conn = open_catalog(str(tmp_path / "c.sqlite"))
    try:
        tables = {
            r[0]
            for r in conn.execute(
                "SELECT name FROM sqlite_master WHERE type='table'"
            )
        }
        assert "files" in tables
        assert "volumes" in tables
        assert "meta" in tables
    finally:
        conn.close()
